File: scratch/test_ws.py
import asyncio
import websockets
import urllib.request
import json
import ssl

async def test_auth_header(token):
    print("Testing with Authorization header...")
    url = "wss://meowtopia-panel.duckdns.org/proxy/daemon/server/946f16b4/console"
    headers = {"Authorization": f"Bearer {token}"}
    try:
        async with websockets.connect(url, extra_headers=headers) as ws:
            print("Successfully connected with Authorization header!")
            # Read one message
            msg = await ws.recv()
            print("Received:", msg)
            return True
    except Exception as e:
        print("Failed with Authorization header:", e)
        return False

async def test_subprotocol(token):
    print("\nTesting with subprotocol ['bearer', token]...")
    url = "wss://meowtopia-panel.duckdns.org/proxy/daemon/server/946f16b4/console"
    try:
        async with websockets.connect(url, subprotocols=["bearer", token]) as ws:
            print("Successfully connected with subprotocols!")
            msg = await ws.recv()
            print("Received:", msg)
            return True
    except Exception as e:
        print("Failed with subprotocols:", e)
        return False

async def test_query_param(token, param_name):
    print(f"\nTesting with query param ?{param_name}=...")
    url = f"wss://meowtopia-panel.duckdns.org/proxy/daemon/server/946f16b4/console?{param_name}={token}"
    try:
        async with websockets.connect(url) as ws:
            print(f"Successfully connected with ?{param_name}!")
            msg = await ws.recv()
            print("Received:", msg)
            return True
    except Exception as e:
        print(f"Failed with ?{param_name}:", e)
        return False

def get_token():
    url = "https://meowtopia-panel.duckdns.org/oauth2/token"
    data = "grant_type=client_credentials&client_id=028831fc-688b-4811-a1f8-ce7008958cb6&client_secret=nel4nBv8VvSnTog0od8Oy-43045LSEn-jO56MRA9g7ngJwq3"
    req = urllib.request.Request(
        url, 
        data=data.encode('utf-8'), 
        headers={"Content-Type": "application/x-www-form-urlencoded"}
    )
    with urllib.request.urlopen(req) as response:
        res = json.loads(response.read().decode())
        return res["access_token"]

async def main():
    token = get_token()
    print("Token fetched successfully.")
    
    # Try all methods
    await test_auth_header(token)
    await test_subprotocol(token)
    await test_query_param(token, "token")
    await test_query_param(token, "access_token")

if __name__ == "__main__":
    asyncio.run(main())

File: scratch/test_ws.js
const WebSocket = require('ws');

async function get_token() {
    const url = "https://meowtopia-panel.duckdns.org/oauth2/token";
    const data = "grant_type=client_credentials&client_id=028831fc-688b-4811-a1f8-ce7008958cb6&client_secret=nel4nBv8VvSnTog0od8Oy-43045LSEn-jO56MRA9g7ngJwq3";
    const res = await fetch(url, {
        method: "POST",
        headers: { "Content-Type": "application/x-www-form-urlencoded" },
        body: data
    });
    const json = await res.json();
    return json.access_token;
}

const paths = [
    "/proxy/daemon/server/946f16b4/console",
    "/api/servers/946f16b4/console",
    "/api/servers/946f16b4/socket",
    "/api/server/946f16b4/console",
    "/api/servers/946f16b4/ws",
];

async function test_path(token, path) {
    console.log(`\n--- Testing path: ${path} ---`);
    
    // Test 1: ws with query param ?token=
    await new Promise((resolve) => {
        const url = `wss://meowtopia-panel.duckdns.org${path}?token=${token}`;
        const ws = new WebSocket(url);
        ws.on('open', () => {
            console.log(`  [OK] Connected via ?token=`);
            ws.close();
            resolve(true);
        });
        ws.on('error', (err) => {
            console.log(`  [FAIL] ?token= failed:`, err.message);
            resolve(false);
        });
    });

    // Test 2: ws with Auth header
    await new Promise((resolve) => {
        const url = `wss://meowtopia-panel.duckdns.org${path}`;
        const ws = new WebSocket(url, {
            headers: { "Authorization": `Bearer ${token}` }
        });
        ws.on('open', () => {
            console.log(`  [OK] Connected via Authorization header`);
            ws.close();
            resolve(true);
        });
        ws.on('error', (err) => {
            console.log(`  [FAIL] Auth header failed:`, err.message);
            resolve(false);
        });
    });

    // Test 3: ws with subprotocol
    await new Promise((resolve) => {
        const url = `wss://meowtopia-panel.duckdns.org${path}`;
        const ws = new WebSocket(url, ["bearer", token]);
        ws.on('open', () => {
            console.log(`  [OK] Connected via subprotocol`);
            ws.close();
            resolve(true);
        });
        ws.on('error', (err) => {
            console.log(`  [FAIL] subprotocol failed:`, err.message);
            resolve(false);
        });
    });
}

async function main() {
    const token = await get_token();
    console.log("Token fetched.");
    for (const p of paths) {
        await test_path(token, p);
    }
}

main().catch(console.error);
